File: web-client/src/api/costs-materials.ts
/**
 * API client for Costs and Materials endpoints
 */
import apiClient from './client'
import type {
  Unit,
  CostItem,
  Material,
  CostItemMaterial,
  WorkComposition
} from '@/types/models'

// ============================================================================
// Units API
// ============================================================================

export const unitsApi = {
  /**
   * Get all units
   */
  async getAll(includeDeleted = false): Promise<Unit[]> {
    const response = await apiClient.get<Unit[]>('/units', {
      params: { include_deleted: includeDeleted }
    })
    return response.data
  },

  /**
   * Get unit by ID
   */
  async getById(id: number): Promise<Unit> {
    const response = await apiClient.get<Unit>(`/units/${id}`)
    return response.data
  },

  /**
   * Create new unit
   */
  async create(data: Omit<Unit, 'id' | 'marked_for_deletion' | 'created_at' | 'modified_at'>): Promise<Unit> {
    const response = await apiClient.post<Unit>('/units', data)
    return response.data
  },

  /**
   * Update unit
   */
  async update(id: number, data: Partial<Unit>): Promise<Unit> {
    const response = await apiClient.put<Unit>(`/units/${id}`, data)
    return response.data
  },

  /**
   * Delete unit
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(`/units/${id}`)
  }
}

// ============================================================================
// Cost Items API
// ============================================================================

export const costItemsApi = {
  /**
   * Get all cost items
   */
  async getAll(includeDeleted = false): Promise<CostItem[]> {
    const response = await apiClient.get<CostItem[]>('/cost-items', {
      params: { include_deleted: includeDeleted }
    })
    return response.data
  },

  /**
   * Get cost item by ID
   */
  async getById(id: number): Promise<CostItem> {
    const response = await apiClient.get<CostItem>(`/cost-items/${id}`)
    return response.data
  },

  /**
   * Create new cost item
   */
  async create(data: Omit<CostItem, 'id' | 'marked_for_deletion' | 'created_at' | 'modified_at'>): Promise<CostItem> {
    const response = await apiClient.post<CostItem>('/cost-items', data)
    return response.data
  },

  /**
   * Update cost item
   */
  async update(id: number, data: Partial<CostItem>): Promise<CostItem> {
    const response = await apiClient.put<CostItem>(`/cost-items/${id}`, data)
    return response.data
  },

  /**
   * Delete cost item
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(`/cost-items/${id}`)
  }
}

// ============================================================================
// Materials API
// ============================================================================

export const materialsApi = {
  /**
   * Get all materials
   */
  async getAll(includeDeleted = false): Promise<Material[]> {
    const response = await apiClient.get<Material[]>('/materials', {
      params: { include_deleted: includeDeleted }
    })
    return response.data
  },

  /**
   * Get material by ID
   */
  async getById(id: number): Promise<Material> {
    const response = await apiClient.get<Material>(`/materials/${id}`)
    return response.data
  },

  /**
   * Create new material
   */
  async create(data: Omit<Material, 'id' | 'marked_for_deletion' | 'created_at' | 'modified_at'>): Promise<Material> {
    const response = await apiClient.post<Material>('/materials', data)
    return response.data
  },

  /**
   * Update material
   */
  async update(id: number, data: Partial<Material>): Promise<Material> {
    const response = await apiClient.put<Material>(`/materials/${id}`, data)
    return response.data
  },

  /**
   * Delete material
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(`/materials/${id}`)
  }
}

// ============================================================================
// Work Composition API
// ============================================================================

export const workCompositionApi = {
  /**
   * Get complete work composition with cost items and materials
   */
  async getComposition(workId: number): Promise<WorkComposition> {
    const response = await apiClient.get<WorkComposition>(`/works/${workId}/composition`)
    return response.data
  },

  /**
   * Add cost item to work
   */
  async addCostItem(workId: number, costItemId: number): Promise<CostItemMaterial> {
    const response = await apiClient.post<CostItemMaterial>(
      `/works/${workId}/cost-items`,
      { cost_item_id: costItemId }
    )
    return response.data
  },

  /**
   * Add material to work
   */
  async addMaterial(
    workId: number,
    data: {
      cost_item_id: number
      material_id: number
      quantity_per_unit: number
    }
  ): Promise<CostItemMaterial> {
    const response = await apiClient.post<CostItemMaterial>(
      `/works/${workId}/materials`,
      { ...data, work_id: workId }
    )
    return response.data
  },

  /**
   * Update material quantity
   */
  async updateMaterialQuantity(
    workId: number,
    associationId: number,
    quantityPerUnit: number
  ): Promise<CostItemMaterial> {
    const response = await apiClient.put<CostItemMaterial>(
      `/works/${workId}/materials/${associationId}`,
      { quantity_per_unit: quantityPerUnit }
    )
    return response.data
  },

  /**
   * Change material cost item association
   */
  async changeMaterialCostItem(
    workId: number,
    associationId: number,
    newCostItemId: number
  ): Promise<CostItemMaterial> {
    const response = await apiClient.put<CostItemMaterial>(
      `/works/${workId}/materials/${associationId}`,
      { cost_item_id: newCostItemId }
    )
    return response.data
  },

  /**
   * Remove cost item from work
   */
  async removeCostItem(workId: number, costItemId: number): Promise<void> {
    await apiClient.delete(`/works/${workId}/cost-items/${costItemId}`)
  },

  /**
   * Remove material from work
   */
  async removeMaterial(workId: number, associationId: number): Promise<void> {
    await apiClient.delete(`/works/${workId}/materials/${associationId}`)
  }
}

File: api/models/costs_materials.py
"""
Cost Items and Materials models for API

This module defines Pydantic models for costs and materials related endpoints.
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


# ============================================================================
# Unit Models
# ============================================================================

class UnitBase(BaseModel):
    """Base model for unit"""
    name: str = Field(..., min_length=1, max_length=50)
    description: Optional[str] = None


class UnitCreate(UnitBase):
    """Model for creating unit"""
    pass


class UnitUpdate(UnitBase):
    """Model for updating unit"""
    pass


class Unit(UnitBase):
    """Model for unit with ID"""
    id: int
    marked_for_deletion: bool = False
    created_at: Optional[datetime] = None
    modified_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# ============================================================================
# Cost Item Models
# ============================================================================

class CostItemBase(BaseModel):
    """Base model for cost item"""
    parent_id: Optional[int] = None
    code: Optional[str] = None
    description: str = Field(..., max_length=500)
    is_folder: bool = False
    price: float = Field(default=0.0, ge=0)
    unit: Optional[str] = None  # Legacy
    unit_id: Optional[int] = None
    labor_coefficient: float = Field(default=0.0, ge=0)


class CostItemCreate(CostItemBase):
    """Model for creating cost item"""
    pass


class CostItemUpdate(CostItemBase):
    """Model for updating cost item"""
    pass


class CostItem(CostItemBase):
    """Model for cost item with ID"""
    id: int
    unit_name: Optional[str] = None  # Joined from units table
    marked_for_deletion: bool = False
    created_at: Optional[datetime] = None
    modified_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# ============================================================================
# Material Models
# ============================================================================

class MaterialBase(BaseModel):
    """Base model for material"""
    code: Optional[str] = None
    description: str = Field(..., max_length=500)
    price: float = Field(default=0.0, ge=0)
    unit: Optional[str] = None  # Legacy
    unit_id: Optional[int] = None


class MaterialCreate(MaterialBase):
    """Model for creating material"""
    pass


class MaterialUpdate(MaterialBase):
    """Model for updating material"""
    pass


class Material(MaterialBase):
    """Model for material with ID"""
    id: int
    unit_name: Optional[str] = None  # Joined from units table
    marked_for_deletion: bool = False
    created_at: Optional[datetime] = None
    modified_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# ============================================================================
# Cost Item Material Association Models
# ============================================================================

class CostItemMaterialBase(BaseModel):
    """Base model for cost item material association"""
    work_id: int
    cost_item_id: int
    material_id: Optional[int] = None
    quantity_per_unit: float = Field(default=0.0, ge=0)


class CostItemMaterialCreate(CostItemMaterialBase):
    """Model for creating cost item material association"""
    pass


class CostItemMaterialUpdate(BaseModel):
    """Model for updating cost item material association"""
    quantity_per_unit: Optional[float] = Field(None, gt=0)
    cost_item_id: Optional[int] = None


class CostItemMaterialSimple(CostItemMaterialBase):
    """Simple model for cost item material association with ID (no nested objects)"""
    id: int
    
    class Config:
        from_attributes = True


class CostItemMaterial(CostItemMaterialBase):
    """Model for cost item material association with ID and nested objects"""
    id: int
    cost_item: Optional[CostItem] = None
    material: Optional[Material] = None
    
    class Config:
        from_attributes = True


# ============================================================================
# Work Composition Models
# ============================================================================

class WorkComposition(BaseModel):
    """Complete work composition with cost items and materials"""
    work_id: int
    work_name: str
    cost_items: List[CostItemMaterial] = []
    materials: List[CostItemMaterial] = []
    total_cost: float = 0.0


class WorkCompositionDetail(BaseModel):
    """Detailed work composition for UI"""
    work_id: int
    work_name: str
    work_code: Optional[str] = None
    work_unit: Optional[str] = None
    work_price: float = 0.0
    work_labor_rate: float = 0.0
    
    # Cost items (where material_id is NULL)
    cost_items: List[CostItemMaterial] = []
    
    # Materials (where material_id is NOT NULL)
    materials: List[CostItemMaterial] = []
    
    # Calculated totals
    total_cost_items_price: float = 0.0
    total_materials_cost: float = 0.0
    total_cost: float = 0.0
